doc/sphinx/source/debug.rst

```diff
@@ -0,0 +1,390 @@
+    6   {
+    7     size_t gid = get_global_id(0);
+    8
+    9     gid += 18298392UL;
+    10    c[gid] = a[gid] * b[gid] + (float4)(1.0f, 6.0f, 9.0f, 4.0f);
+    11  }
+    (gdb) print gid
+    $1 = 18298392
+    (gdb) bt
+    #0  0x00007fffec532458 in dot_product (a=0x5555557bb580, b=0x5555557e6500, c=0x5555557ba480) at /tmp/POCL_CACHE/tempfile-1c-aa-cd-3e-5e.cl:10
+    #1  0x00007fffec5324c3 in _pocl_kernel_dot_product_workgroup ()
+       from /tmp/POCL_CACHE/BJ/JMEICBEBICMMDJCKNIADBFKHIMHDBIIKHCHED/dot_product/2-1-1-goffs0-smallgrid/dot_product.so
+    #2  0x00007ffff72924ed in work_group_scheduler (k=0x7fffb91935c0, thread_data=0x5555557ae600)
+        at /tmp/pocl_source/lib/CL/devices/pthread/pthread_scheduler.c:307
+    #3  0x00007ffff7292b72 in pthread_scheduler_get_work (td=0x5555557ae600) at /tmp/pocl_source/lib/CL/devices/pthread/pthread_scheduler.c:454
+    #4  0x00007ffff7292fd2 in pocl_pthread_driver_thread (p=0x5555557ae600) at /tmp/pocl_source/lib/CL/devices/pthread/pthread_scheduler.c:530
+    #5  0x00007fffee90e6db in start_thread (arg=0x7fffddffe700) at pthread_create.c:463
+    #6  0x00007ffff78faa3f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
+
+Example 2:
+
+Lets say we want to step the kernel from previous example. Launch gdb::
+
+    POCL_MAX_PTHREAD_COUNT=1 gdb ./example
+
+We can't set a breakpoint inside a kernel before running the program,
+because PoCL writes the source it receives at runtime
+(by ``clCreateProgramWithSource``) to a temporary file.
+
+First we need to find some place in the pthread driver sources
+where the kernel is fully compiled but not yet executed.
+A good place is where the pthread driver launches the
+workgroup function. In the pthread driver, this is a call to
+``(struct kernel_run_command)->workgroup``, looking like this::
+
+	  k->workgroup ((uint8_t*)arguments, (uint8_t*)&pc,
+		gids[0], gids[1], gids[2]);
+
+Lets say it's at ``lib/CL/devices/pthread/pthread_scheduler.c:307``, set the breakpoint::
+
+    (gdb) break /tmp/pocl_source/lib/CL/devices/pthread/pthread_scheduler.c:307
+	No source file named /tmp/pocl_source/lib/CL/devices/pthread/pthread_scheduler.c.
+	Make breakpoint pending on future shared library load? (y or [n]) y
+	Breakpoint 1 (/tmp/pocl_source/lib/CL/devices/pthread/pthread_scheduler.c:307) pending.
+
+Run the program::
+
+	(gdb) r
+	Starting program: /tmp/pocl_build/example
+	[Thread debugging using libthread_db enabled]
+	Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
+	[New Thread 0x7fffedf36700 (LWP 11851)]
+	[Switching to Thread 0x7fffedf36700 (LWP 11851)]
+
+	Thread 2 "example" hit Breakpoint 1, work_group_scheduler (k=0x7fffe9001640, thread_data=0x5555557ad500)
+		at /tmp/pocl_source/lib/CL/devices/pthread/pthread_scheduler.c:307
+	307	          k->workgroup ((uint8_t*)arguments, (uint8_t*)&pc,
+
+
+Now it's about to launch the ``dot_product`` kernel for a single workgroup, so let's find the kernel source file::
```

Comment:
Phew. Quite a bit of hoops to go through. Why not just make the POCL_DEBUG dump the temp file name more explicitly so it can be used in gdb?

Reply:
We can do that, but it won't change the fact that you have to stop at `k->workgroup`  (or somewhere post-compilation-before-execution). So it would really only save one step, the `info functions kernel_name`. If you think it's worth it, i can add the code.

Reply:
Can we use the kernel function name to set the breakpoint (including the PoCL WG naming)?

Reply:
Yes, that works, and it makes the process significantly simpler. I've updated the docs